aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu

```diff
@@ -0,0 +1,517 @@
+  // gradOutput offset by slice/feature and frame/time
+  scalar_t *gradOutput_nt = gradOutput + o_plane*osizeH*osizeW;
+
+  // For all output pixels...
+  for (oh = ostartH; oh < oendH; oh += ostepH) {
+    int istartH = start_index(oh, osizeH, isizeH);
+    int iendH = end_index(oh, osizeH, isizeH);
+    int kH = iendH - istartH;
+
+    for (ow = ostartW; ow < oendW; ow += ostepW) {
+      int istartW = start_index(ow, osizeW, isizeW);
+      int iendW = end_index(ow, osizeW, isizeW);
+      int kW = iendW - istartW;
+
+      // Compute the gradients from corresponding input pixels
+      scalar_t *ptr_gradInput = gradInput_nt + istartH*isizeW + istartW;
+      scalar_t *ptr_gradOutput = gradOutput_nt + oh*osizeW + ow;
+      scalar_t grad_delta = *ptr_gradOutput / kT / kH / kW;
+
+      int it, ih, iw;
+      for (it = 0; it < kT; ++it) {
+        for (ih = 0; ih < kH; ++ih) {
+          for (iw = 0; iw < kW; ++iw) {
+            atomicAdd(&(ptr_gradInput[ih*isizeW + iw]), grad_delta);
+          }
+        }
+        ptr_gradInput += isizeH*isizeW; // next input frame
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+void atomicadaptiveaveragegradinput_loop(
+    scalar_t* gradInput_data, scalar_t* gradOutput_data,
+    int64_t totalZ,
+    int isizeT, int isizeH, int isizeW,
+    int osizeT, int osizeH, int osizeW) {
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  int blocksH = std::max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
+    atomicadaptiveaveragegradinput<<<blocks, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
+        gradInput_data, gradOutput_data,
+        isizeT, isizeH, isizeW,
+        osizeT, osizeH, osizeW,
+        offsetZ);
+
+    totalZ -= 65535;
+    offsetZ += 65535;
+    AT_CUDA_CHECK(cudaGetLastError());
+  }
+}
+
+// 5D tensor B x D x T x H x w
+
+void adaptive_avg_pool3d_out_cuda_template(
+    Tensor& output,
+    const Tensor& input_,
+    IntArrayRef& output_size) {
+  TensorArg output_arg{output, "output", 1};
+  TensorArg input_arg{input_, "input_", 2};
+
+  checkAllSameGPU("adaptive_avg_pool3d_cuda", {output_arg, input_arg});
+
+  for (int64_t i = 0; i < input_.ndimension(); i++) {
+    AT_CHECK(
+        input_.size(i) > 0,
+        "adaptive_avg_pool3d_cuda(): expected input to have non-empty spatial dimensions, "
+        "but input has sizes ", input_.sizes(),
+        " with dimension ", i, " being empty");
+  }
+
+  AT_CHECK(
+      (input_.ndimension() == 4 || input_.ndimension() == 5),
+      "non-empty 4D or 5D (batch mode) tensor expected for input");
+
+  // the jit sometimes passes output_size.size() == 1
+  AT_CHECK(
+      output_size.size() == 1 || output_size.size() == 3,
+      "adaptive_avg_pool3d: internal error: output_size.size() must be 1 or 3");
+
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size[1];
+  int64_t osizeW = output_size[2];
+
```

Comment:
If you allow `output.size.size() == 1`, you also have to set `osizeH` and `osizeW` conditionally.

But I think the `JIT` issue does not happen for `AdaptiveAveragePooling`, so just checking for `output_size.size() == 3` may be sufficient if the tests pass.

Reply:
Can we file an issue about this, and reference it to all of the cases where we added extra checks like this? I'm pretty sure this code is supposed to not be necessary, but since no one is root causing it, it looks like this code is going to replicate everywhere, can we at least put a comment on it.

Reply:
Yes, it's a good idea to file an issue, I'll open one.  Probably for this file the workaround is not necessary, but I haven't verified.